src/ts_generator.py

```diff
@@ -0,0 +1,364 @@
+                    B[day] = 0
+
+    def duration_generator(
+        self, law: ProbilityLaw, volatility: float, a: float, b: float, expec: int
+    ) -> int:
+        """
+        generation of random outage duration
+        """
+        rnd_nb = rndgenerator.next()
+        if law == ProbilityLaw.UNIFORM:
+            return int(a + rnd_nb * b)
+        elif law == ProbilityLaw.GEOMETRIC:
+            return min(int(1 + a + b * log(rnd_nb)), int(self.log_size / 2 - 1))
+
+    def generate_time_series(
+        self,
+        cluster: ThermalCluster,
+        number_of_timeseries: int,
+        output_series: List[List[float]],
+        output_outages: Optional[List[List[Tuple[int, int, int, int, int]]]] = None,
+    ) -> None:
+        """
+        generation of multiple timeseries for a given thermal cluster
+        output_series stores available power at a given time
+        output_outages stores infomation about outages begining at a given time -> (PPO, PFO, MXO, POD, FOD)
+        """
+
+        # --- precalculation ---
+        # cached values for (1-lf)**k and (1-lp)**k
+        self.FPOW: List[List[float]] = []
+        self.PPOW: List[List[float]] = []
+
+        for day in range(self.days_per_year):
+            # lf and lp represent the forced and programed failure rate
+            # failure rate means the probability to enter in outage each day
+            # its value is given by: OR / [OR + OD * (1 - OR)]
+            FOR = cluster.fo_rate[day]
+            FOD = cluster.fo_duration[day]
+            self.lf[day] = FOR / (FOR + FOD * (1 - FOR))
+
+            POR = cluster.po_rate[day]
+            POD = cluster.po_duration[day]
+            self.lp[day] = POR / (POR + POD * (1 - POR))
+
+            if self.lf[day] < 0:
+                raise ValueError(f"forced failure rate is negative on day {day}")
+            if self.lp[day] < 0:
+                raise ValueError(f"programed failure rate is negative on day {day}")
+
+            ## i dont understand what these calulations are for
+            ## consequently reduce the lower failure rate
+            if self.lf[day] < self.lp[day]:
+                self.lf[day] *= (1 - self.lp[day]) / (1 - self.lf[day])
+            if self.lp[day] < self.lf[day]:
+                self.lp[day] *= (1 - self.lf[day]) / (1 - self.lp[day])
+
+            a = 0
+            b = 0
+            if self.lf[day] <= FAILURE_RATE_EQ_1:
+                a = 1 - self.lf[day]
+                self.ff[day] = self.lf[day] / a
+            if self.lp[day] <= FAILURE_RATE_EQ_1:
+                b = 1 - self.lp[day]
+                self.pp[day] = self.lp[day] / b
+
+            # pre calculating power values
+            self.FPOW.append([])
+            self.PPOW.append([])
+            for k in range(cluster.unit_count + 1):
+                self.FPOW[-1].append(pow(a, k))
+                self.PPOW[-1].append(pow(b, k))
+
+        self.prepare_outage_duration_constant(
```

Comment:
First idea to start improving readability of the process: it seems like the duration generation is a rather isolated process. We could push further the separation with `ThermalDataGenerator` by moving the related methods and the related parameters (A, B) into another class `DurationGenerator`.

We could actually have a small hierarchy to implement uniform law and geometric low in separate classes, like `UniformLawDurationGenetor` and `̀GeometricLAwDurationGenerator` that would implement the abstract class `DurationGenerator`.